pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -167,6 +167,13 @@ func (builder *ListenerBuilder) aggregateVirtualInboundListener() *ListenerBuild
 	// If TLS inspector sets transport protocol to tls, the http inspector
 	// won't inspect the packet.
 	if util.IsProtocolSniffingEnabledForInbound(builder.node) {
+		if !needTLS {
```

Comment:
interesting, why need to add tls inspector when tls is not needed?

Reply:
I think it's a bug in envoy where if alpn is set it requires tls inspector. but really that's not correct. Can we just fix envoy?

Reply:
From here, i think you should fix the reduceInboundListenerToFilterChains
 ```
  const bool need_tls_inspector =
      std::any_of(
          config.filter_chains().begin(), config.filter_chains().end(),
          [](const auto& filter_chain) {
            const auto& matcher = filter_chain.filter_chain_match();
            return matcher.transport_protocol() == "tls" ||
                   (matcher.transport_protocol().empty() &&
                    (!matcher.server_names().empty() || !matcher.application_protocols().empty()));
          }) &&
      !std::any_of(config.listener_filters().begin(), config.listener_filters().end(),
                   [](const auto& filter) {
                     return filter.name() ==
                            Extensions::ListenerFilters::ListenerFilterNames::get().TlsInspector;
                   });
```

Reply:
yep, exactly. Better yet remove the entire code...It's completely against everything else envoy does to automagically fix broken xds config. That may be harder to get merged though

Reply:
The logic is not right in Envoy. But the fix will introduce the http_inspector to the core Envoy code. I don't think it will be easy to get merged. 

The right is logic is:

1. If transport protocol is "tls" or server name is not empty, we should use tls_inspector
2. If application protocol is not empty, we should use both tls_inspector and http_inspector. 

which implies that http_inspector should coexists with tls_inspector. 

According to the above conclusion, it's reasonable to add tls_inspector where http_inspector exists.

Reply:
why do I need tls inspector if I just want to sniff plaintext http?

Reply:
You don't know if it's for plaintext or not.

Reply:
If the application protocol is set, it could come from tls or plaintext.

Reply:
Consider this scenario. I send config:
```
alpn: [http]
filters: http_inspector
```

Because I am the control plane, I know what I am doing. I do NOT want tls inspector, I just want to tell if its TCP or HTTP. Why should I be forced to add tls_inspector?

I cannot think of a single other place where Envoy enforces arbitrary requirements like this

Reply:
> If the application protocol is set, it could come from tls or plaintext.

If I know, with certainty, that all inbound traffic is plain text, then I also know that the alpn is set by http_inspector, right?

Reply:
> I just want to tell if its TCP or HTTP. Why should I be forced to add tls_inspector?

tls_inspector could also separate TCP from HTTP because tls inspector can set the alpn to http. From Envoy's perspective, it doesn't know the connection is plaintext or not. That's why it enforces the tls inspector.

Reply:
> tls_inspector could also separate TCP from HTTP because tls inspector can set the alpn to http. From Envoy's perspective, it doesn't know the connection is plaintext or not. That's why it enforces the tls inspector.

Why do I NEED tls_inspector? Just because tls_inspector CAN set the alpn doesn't mean I want it to?

Reply:
> If I know, with certainty, that all inbound traffic is plain text, then I also know that the alpn is set by http_inspector, right?

We don't know all inbound traffic is plain text right?

Reply:
> Why do I NEED tls_inspector? Just because tls_inspector CAN set the alpn doesn't mean I want it to?

If you don't want the inspector to set the alpn, you should set the application protocol to empty. If the application protocol is not empty, the alpn could be set by either http inspector or tls inspector.

Reply:
> We don't know all inbound traffic is plain text right?

This is Envoy, not Istio. I can use Envoy on my toaster oven where I know all traffic is plain text if I choose to. We shouldn't put Istio-ism into Envoy

> If you don't want the inspector to set the alpn, you should set the application protocol to empty. If the application protocol is not empty, the alpn could be set by either http inspector or tls inspector.

I think there is some confusion here. I DO want the http_inspector to set the alpn, I do not want tls inspector to set the alpn. Why is that not possible?

Reply:
> I think there is some confusion here. I DO want the http_inspector to set the alpn, I do not want tls inspector to set the alpn. Why is that not possible?

1. We all agree that Envoy doesn't know the connect is plain text or not
2. When Envoy see application protocol is not empty, the Envoy thinks that one of the inspectors set the alpn
3. Even though Envoy sees there is already a http inspector there, it will add another one according to the first statement.

> This is Envoy, not Istio. I can use Envoy on my toaster oven where I know all traffic is plain text if I choose to. We shouldn't put Istio-ism into Envoy

We should remove that logic from Envoy but it's not easy to get merged and also there might be some back comp issue.

Reply:
Can we at least open up an issue tracking this in Envoy side before we merge a hack into istio to bypass it?

Reply:
Read another thread about the tls inspector discussion https://github.com/istio/istio/issues/7601#issuecomment-410831541.

It seems that we cannot remove that check in Envoy.

Reply:
That is just about the log. I am suggesting we do not even inject it at all

Reply:
@PiotrSikora what do you think about remove the check for tls inspector code from the Envoy. The control plane should be responsible for the right configuration.

https://github.com/envoyproxy/envoy/blob/727a1c1bc33a38354ccc0dd27ba735bb36cddfc2/source/server/listener_impl.cc#L275

It causes some mess when application protocol is used. Envoy should inject both http inspector and tls inspector.

Reply:
> That is just about the log. I am suggesting we do not even inject it at all

Another reason to keep the check in Envoy would be for backward compatibility.

Reply:
@howardjohn if you know that all traffic is plaintext, then you can set `transport_protocol: raw_buffer` and TLS inspector won't be injected.

Reply:
I'd let the control plane be explicit here. Any guessing done by proxy is going to cause problems when we need to migrate to a different configuration model. The root issue is that declarative filter chain condition is not trivially executable since you can't inspect both TLS and HTTP without decryption. So you would need to sequence operations to inspect TLS, decrypt TLS, inspect HTTP, parse HTTP, etc. This is a non-trivial program so should be done by the control plane.